libraries/AC_PID/AC_PID_1D.cpp

```diff
@@ -0,0 +1,248 @@
+        _error += get_filt_E_alpha() * ((_target - measurement) - _error);
+
+        // calculate and filter derivative
+        if (_dt > 0.0f) {
+            float derivative = (_error - error_last) / _dt;
+            _derivative += get_filt_D_alpha() * (derivative - _derivative);
+        }
+    }
+
+    // update I term
+    update_i(limit_neg, limit_pos);
+
+    float P_out = (_error * _kp);
+    float D_out = (_derivative * _kd);
+
+    _pid_info.target = _target;
+    _pid_info.actual = measurement;
+    _pid_info.error = _error;
+    _pid_info.P = _error * _kp;
+    _pid_info.I = _integrator;
+    _pid_info.D = _derivative * _kd;
+    _pid_info.FF = _target * _kff;
+
+    return P_out + _integrator + D_out + _target * _kff;
+}
+
+//  update_i - update the integral
+//  If the limit flag is set the integral is only allowed to shrink
+void AC_PID_1D::update_i(bool limit_neg, bool limit_pos)
+{
+    if (!is_zero(_ki) && is_positive(_dt)) {
+        // Ensure that integrator can only be reduced if the output is saturated
+        if (!((limit_neg && is_negative(_error)) || (limit_pos && is_positive(_error)))) {
+            _integrator += ((float)_error * _ki) * _dt;
+            _integrator = constrain_float(_integrator, -_kimax, _kimax);
+        }
+    } else {
+        _integrator = 0.0f;
+    }
+}
+
+float AC_PID_1D::get_p() const
+{
+    return _error * _kp;
+}
+
+float AC_PID_1D::get_i() const
+{
+    return _integrator;
+}
+
+float AC_PID_1D::get_d() const
+{
+    return _derivative * _kd;
+}
+
+float AC_PID_1D::get_ff()
+{
+    return _target * _kff;
+}
+
+void AC_PID_1D::reset_I()
+{
+    _integrator = 0.0f;
+}
+
+void AC_PID_1D::load_gains()
+{
+    _kp.load();
+    _ki.load();
+    _kd.load();
+    _kff.load();
+    _kimax.load();
+    _kimax = fabsf(_kimax);
+    _filt_E_hz.load();
+    _filt_D_hz.load();
+}
+
+// save_gains - save gains to eeprom
+void AC_PID_1D::save_gains()
+{
+    _kp.save();
+    _ki.save();
+    _kd.save();
+    _kff.save();
+    _kimax.save();
+    _filt_E_hz.save();
+    _filt_D_hz.save();
+}
+
+/// Overload the function call operator to permit easy initialisation
+void AC_PID_1D::operator()(float p_val, float i_val, float d_val, float ff_val, float imax_val, float input_filt_E_hz, float input_filt_D_hz, float dt)
+{
+    _kp = p_val;
+    _ki = i_val;
+    _kd = d_val;
+    _kff = ff_val;
+    _kimax = fabsf(imax_val);
+    _filt_E_hz = input_filt_E_hz;
+    _filt_D_hz = input_filt_D_hz;
+    _dt = dt;
+}
+
+// get_filt_T_alpha - get the target filter alpha
+float AC_PID_1D::get_filt_E_alpha() const
+{
+    return get_filt_alpha(_filt_E_hz);
+}
+
+// get_filt_D_alpha - get the derivative filter alpha
+float AC_PID_1D::get_filt_D_alpha() const
+{
+    return get_filt_alpha(_filt_D_hz);
+}
+
+// get_filt_alpha - calculate a filter alpha
+float AC_PID_1D::get_filt_alpha(float filt_hz) const
+{
+    if (is_zero(filt_hz)) {
+        return 1.0f;
+    }
+
+    // calculate alpha
```

Comment:
Didn't tridge factor this out somewhere recently?

Reply:
ok, changed AC_PID_1D to use calc_lowpass_alpha_dt